skbio/tree/_dists.py

```diff
@@ -0,0 +1,395 @@
+    shuffler : int, np.random.Generator or callable, optional
+        The shuffling function to use if ``sample`` is specified. Default is
+        :meth:`~numpy.random.Generator.shuffle`. If an integer is provided, a random
+        generator will be constructed using this number as the seed.
+
+    Returns
+    -------
+    DistanceMatrix
+        Matrix of the path-length distances or variants.
+
+    See Also
+    --------
+    TreeNode.compare_cophenet
+
+    Notes
+    -----
+    The path-length distance [1]_ is the square root of the sum of squared differences
+    of path lengths among all pairs of taxa between two trees.
+
+    This function is equivalent to :meth:`TreeNode.compare_cophenet` for two trees.
+    Refer to the latter for details about the metric and its variants, and parameter
+    settings for calculating them. However, the current function extends the operation
+    to an arbitrary number of trees and returns a distance matrix for them. It is named
+    so because the term "cophenetic distance" refers to the distance between two taxa
+    in a tree instead.
+
+    A restriction of the current function compared to ``compare_cophenet`` is that
+    ``metric`` must be symmetric (i.e., :math:`d(x, y) = d(y, x)`), and equals zero
+    from a vector to itself (i.e., :math:`d(x, x) = 0`). It does not have to suffice
+    non-negativity or triangle inequality though.
+
+    This function is optimized for calculation based on taxa shared across all trees.
+    One can instead set ``pairwise`` to True to calculate based on taxa shared between
+    each pair of trees, which is however less efficient as the path lengths need to be
+    re-calculated during each comparison.
+
+    References
+    ----------
+    .. [1] Lapointe, F. J., & Cucumel, G. (1997). The average consensus procedure:
+        combination of weighted trees containing identical or overlapping sets of
+        taxa. Systematic Biology, 46(2), 306-312.
+
+    Examples
+    --------
+    >>> trees = [TreeNode.read([x]) for x in (
+    ...     "((a:1,b:2):1,c:4,((d:4,e:5):2,f:6):1);",
+    ...     "((a:3,(b:2,c:2):1):3,d:8,(e:5,f:6):2);",
+    ...     "((a:1,c:6):2,(b:3,(d:2,e:3):1):2,f:7);",
+    ... )]
+    >>> dm = path_dists(trees, ids=list("ABC"))
+    >>> print(dm)
+    3x3 distance matrix
+    IDs:
+    'A', 'B', 'C'
+    Data:
+    [[  0.          13.7113092   11.87434209]
+     [ 13.7113092    0.          19.5192213 ]
+     [ 11.87434209  19.5192213    0.        ]]
+
+    """
+    _check_ids(trees, ids)
+
+    if sample is not None:
+        shuffler = _check_shuffler(shuffler)
+
+    if pairwise:
+        metric = partial(
+            TreeNode.compare_cophenet,
+            sample=sample,
+            metric=metric,
+            shuffler=shuffler,
+            use_length=use_length,
+            ignore_self=True,
+        )
+        return DistanceMatrix.from_iterable(
+            trees, metric=metric, keys=ids, validate=False
```

Comment:
`metric` here I _think_ would be the `repr` of the `partial` function from line 358 which is likely not what's intended

Reply:
FWIW, I've recently shifted to using PyRight as an LSP, alongside ruff. It would be painful to adapt in skbio because it actually uses typing within the IDE, and I think would require we move to type the project as a whole. But, it is very good at catching these types of issues, among others, so there may be long term benefit.

Reply:
Great thought! I guess that typing can facilitate JIT by future Python interpreters. Since we ditch Python 3.8 now, I think that Python 3.9+ has better built-in support for type hinting. One the other hand, enforcing typing will require lots of efforts though, and some may not be straightforward (therefore cannot be directly outsourced to an undergrad). SciPy and scikit-learn haven't adopted typing, but pandas has.